### PR TITLE
Change the structure of how we save stores

### PR DESCRIPTION
**Pull request by GustavoCaso**
We move from an Array to a Hash, now we can access the stores using a
hash notation.

Validate when a new store is added, if the store has not the methods
defined we will raise an exception.

THIS IS A BREAKING CHANGE I will be bumping the version to `1.0.0`

### CODE REVIEW

File: test/progressrus_test.rb [hunk 1]
```diff
@@ -234,47 +242,32 @@ def test_date_fields_should_deserialize_properly
...
+    Progressrus.scope(["oemg"])
   end
 
   def test_support_scope_by_name
-    Progressrus.stores.clear
+    Progressrus.clear_stores
 
-    mysql = mock()
-    redis = mock()
+    mysql = mock_shop
```
Comment (moechaieb) [inline test/progressrus_test.rb]: I think you mean `mock_store`?

Comment (GustavoCaso) [reply test/progressrus_test.rb]: Shopify is in my veins 😆

File: lib/progressrus.rb [hunk 2]
```diff
@@ -2,29 +2,60 @@
...
+      super(message)
+    end
+  end
+
   class << self
+    def clear_stores
+      @stores = {}
+    end
+
     def stores
```
Comment (kirs) [inline lib/progressrus.rb]: I was going to say that we might want to use a mutex so this memoization doesn't become thread unsafe for someone who uses this gem with Puma or Sidekiq - but this is completely outside of the scope if your changes.

File: lib/progressrus.rb [hunk 3]
```diff
@@ -2,29 +2,60 @@
...
+
+    def scope(scope, store: :redis)
+      stores[store].scope(scope)
     end
     alias_method :all, :scope
 
-    def find(scope, id, store: :first)
-      stores.find_by_name(store).find(scope, id)
+    def find(scope, id, store: :redis)
+      stores[store].find(scope, id)
```
Comment (kirs) [inline lib/progressrus.rb]: 👍

File: lib/progressrus.rb [hunk 4]
```diff
@@ -2,29 +2,60 @@
...
+
+    def flush(scope, id = nil, store: :redis)
+      stores[store].flush(scope, id)
     end
 
-    def flush(scope, id = nil, store: :first)
-      stores.find_by_name(store).flush(scope, id)
+    private
+
+    def validate_store!(store)
```
Comment (kirs) [inline lib/progressrus.rb]: Is it really necessary to validate those?
At least for adapter-like things in the Rails codebase, I haven't seen this pattern of iterating over methods and checking that it implements all required things. This feels a bit against Ruby.

If someone didn't implement `find` on the adapter, you just let them crash next time they call `find` on Progressrus.

File: lib/progressrus.rb [hunk 5]
```diff
@@ -84,7 +115,7 @@ def complete(now: Time.now)
   end
 
   def flush
-    stores.each { |store| store.flush(scope, id) }
```
Comment (kirs) [inline lib/progressrus.rb]: Dunno how the performance compares, but you could also use `each_value` and avoid discarding `_store_name`. Feel free to ignore this comment :)

File: lib/progressrus.rb [hunk 6]
```diff
@@ -2,29 +2,60 @@
...
+      @stores ||= {
+        redis: Store::Redis.new(::Redis.new(host: ENV["PROGRESSRUS_REDIS_HOST"] || "localhost"))
+      }
     end
 
-    def scope(scope, store: :first)
-      stores.find_by_name(store).scope(scope)
+    def add_store(name, store)
+      validate_store!(store)
+      new_stores = stores.merge(name => store)
```
Comment (kirs) [inline lib/progressrus.rb]: Why `merge` if you could do `stores[name] = store`?
Mutating is not always good, but in this case I don't see any reason not to do it.
`Progressrus.stores` is ok to be mutated directly.

File: lib/progressrus/store.rb [hunk 7]
```diff
@@ -1,24 +0,0 @@
-class Progressrus
-  class Store < Array
```
Comment (kirs) [inline lib/progressrus/store.rb]: 🙈 https://words.steveklabnik.com/beware-subclassing-ruby-core-classes

File: lib/progressrus.rb (hunk 4 above)
Comment (GustavoCaso) [reply lib/progressrus.rb]: I'm ok removing it, just wanted to avoid unwanted errors to users of the library.

BTW `ActiveJob` uses this pattern 
https://github.com/rails/rails/blob/382a9563a3bc3dd1982f4db1258e8f0e7f5fac8e/activejob/lib/active_job/queue_adapter.rb#L50-L59
